editor/src/components/editor/persistence/persistence.ts

```diff
@@ -0,0 +1,231 @@
+              event.projectModel.name,
+              event.projectModel.content,
+            )
+            break
+          case 'PROJECT_ID_CREATED':
+            this.queuedActions.push(setProjectID(event.projectId))
+            break
+          case 'LOAD_FAILED':
+            onProjectNotFound()
+            break
+          case 'DOWNLOAD_ASSETS_COMPLETE': {
+            if (state.matches({ core: { [Forking]: CreatingProjectId } })) {
+              this.queuedActions.push(setForkedFromProjectID(state.context.projectId!))
+              this.queuedActions.push(setProjectName(`${state.context.project!.name} (forked)`))
+              this.queuedActions.push(showToast(notice('Project successfully forked!')))
+            }
+
+            const updateFileActions = event.downloadAssetsResult.filesWithFileNames.map(
+              ({ fileName, file }) => updateFile(fileName, file, true),
+            )
+            this.queuedActions.push(...updateFileActions)
+            break
+          }
+          case 'SAVE_COMPLETE':
+            const updateFileActions = event.saveResult.filesWithFileNames.map(
+              ({ fileName, file }) => updateFile(fileName, file, true),
+            )
+            this.queuedActions.push(...updateFileActions)
+            this.lastSavedTS = Date.now()
+            // TODO Show toasts after first server save of a previously local project
+            break
+        }
+
+        if (state.matches({ core: Ready })) {
+          if (this.queuedActions.length > 0) {
+            const actionsToDispatch = this.queuedActions
+            const projectIdOrNameChanged = actionsToDispatch.some(
+              (action) =>
+                action.action === 'SET_PROJECT_ID' || action.action === 'SET_PROJECT_NAME',
+            )
+            if (projectIdOrNameChanged) {
+              pushProjectURLToBrowserHistory(state.context.projectId!, state.context.project!.name)
+            }
+            this.queuedActions = []
+            dispatch(actionsToDispatch)
+          }
+        }
+      }
+    })
+
+    this.interpreter.start()
+
+    window.addEventListener('beforeunload', (e) => {
+      if (!this.isSafeToClose()) {
+        this.sendThrottledSave()
+        e.preventDefault()
+        e.returnValue = ''
+      }
+    })
+  }
+
+  private isSafeToClose(): boolean {
+    return this.waitingThrottledSaveEvent == null
+  }
+
+  private clearThrottledSave(): void {
+    if (this.throttledSaveTimeoutId != null) {
+      clearTimeout(this.throttledSaveTimeoutId)
+      this.throttledSaveTimeoutId = null
+    }
+
+    this.waitingThrottledSaveEvent = null
+  }
+
+  private getRemainingSaveDelay(): number {
+    return Math.max(0, this.lastSavedTS + this.saveThrottle - Date.now())
+  }
+
+  private shouldThrottle(forceOrThrottle: 'force' | 'throttle'): boolean {
+    return forceOrThrottle === 'throttle' && this.getRemainingSaveDelay() > 0
+  }
+
+  // Exposed for testing
+  sendThrottledSave(): void {
```

Comment:
I think all of these functions should be replaced to arrow functions to avoid accidentally messing up where the `this` points to. I actually thought we had an eslint rule preventing this style!

Reply:
Done :+1: